lib/connections/structs.go

```diff
@@ -165,3 +166,19 @@ func (o *onAddressesChangedNotifier) notifyAddressesChanged(l genericListener) {
 		callback(l)
 	}
 }
+
+type dialTarget struct {
+	dialer   genericDialer
+	uri      *url.URL
+	priority int
+}
```

Comment:
It seems to me that the rationale between what goes into the struct and what goes as parameters is odd - the stuff that is the same for every invocation is in the parameters, and the stuff that varies is in the struct. I realize that this is because we're collecting the stuff that varies into the struct in a loop, and the rest we just "know" and throw in as parameters, but it feels ugly...

It would be nice if the `uri` was the parameter and the rest is the context the dialer works on, making it more liek a normal dialer... I don't have a concrete suggestion on sorting that out though so take it as a suggest/comment and do with it what you like.

I also think it's not worth passing the wait group just for a single done call, it could be done less magically by the caller as `go func() { thedial ; wg.Done() }`

Reply:
How about I wedge everything in the struct, and do the dial in an anonymous function?